python/rascaline/rascaline/utils/clebsch_gordan/_utils.py

```diff
@@ -570,51 +436,26 @@ def cg_tensor_product_blocks_same_samples(
     """
     For a given pair of TensorBlocks and desired angular channels, combines the values
     arrays and returns a new TensorBlock.
+
+    If the samples of the two blocks are different, the intersection samples dimensions
+    are matched and used to reshape the block with fewer dimensions such that the tensor
+    product can be performed.
     """
+    # Reshape the blocks along the samples axis by matching samples dimensions.
+    # TODO: can any reshaping be done on the TensorMap level instead of repeated every
```

Comment:
My worry was that you don't know the specific order of samples until forward is called, so the matching needs to be computed for every pair of blocks anyway, no?

I suppose one could store sample ordering per "center_type", but this would require 1) knowledge of the name of the "center_type" (or "first_atom_type" and "second_atom_type") key is and its specific ordering, and b) sorting of all blocks beforehand to ensure the samples for a given type are in the same order.

Maybe not something worth pursuing now, job for another day?

Reply:
oh yeah, this is for later!